lib/auth/grpcserver.go

```diff
@@ -1065,6 +1066,271 @@ func (g *GRPCServer) DeleteAllKubeServices(ctx context.Context, req *proto.Delet
 	return &empty.Empty{}, nil
 }
 
+func (g *GRPCServer) AddMFADevice(stream proto.AuthService_AddMFADeviceServer) error {
+	ctx := stream.Context()
+	actx, err := g.authenticate(ctx)
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	auth := actx.authServer
+	user := actx.User.GetName()
+
+	// The RPC is streaming both ways and the message sequence is:
+	// (-> means client-to-server, <- means server-to-client)
+	//
+	// 1. -> Init
+	// 2. <- ExistingMFAChallenge
+	// 3. -> ExistingMFAResponse
+	// 4. <- NewMFARegisterChallenge
+	// 5. -> NewMFARegisterResponse
+	// 6. <- Ack
+
+	// 1. receive client Init
+	req, err := stream.Recv()
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	initReq := req.GetInit()
+	if initReq == nil {
+		return trail.ToGRPC(trace.BadParameter("expected AddMFADeviceRequestInit, got %T", req))
+	}
+	devs, err := auth.GetMFADevices(ctx, user)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, d := range devs {
+		if d.Metadata.Name == initReq.DeviceName {
+			return trail.ToGRPC(trace.AlreadyExists("MFA device named %q already exists", d.Metadata.Name))
+		}
+	}
+
+	// 2. send MFAAuthenticateChallenge
+	// Note: authChallenge may be empty if this user has no existing MFA devices.
+	authChallenge, err := auth.mfaAuthChallenge(ctx, user)
```

Comment:
This will still store u2f challenges per-user in the backend (between `mfaAuthChallenge` and `validateMFAAuthResponse`) iiuc which could be avoided by keeping the challenge local to this API?
Same for the registration flow between `RegisterInit` and `RgeisterVerify`.

Overall, it feels that the method could be split into multiple bits that would describe the RPC flow more vividly than the comments.

Reply:
Left TODOs to keep the challenges in memory, will clean this up later.

Good point on breaking up the code, done.